test/jdk/jdk/internal/misc/TerminatingThreadLocal/TestTerminatingThreadLocal.java

```diff
@@ -111,77 +118,57 @@ protected T initialValue() {
         };
 
-        var lock = new Lock();
-
-        var blockerThread = Thread.startVirtualThread(() -> {
-            // force compensation in carrier thread pool which will spin another
-            // carrier thread so that we can later observe it being terminated...
-            synchronized (lock) {
-                while (!lock.unblock) {
-                    try {
-                        lock.wait();
-                    } catch (InterruptedException e) {
-                        throw new RuntimeException(e);
-                    }
-                }
-            }
-            // keep thread running in a non-blocking-fashion which keeps
-            // it bound to carrier thread
-            while (!lock.unspin) {
-                Thread.onSpinWait();
-            }
-        });
-
-        Thread thread = Thread
-            .ofVirtual()
-            .allowSetThreadLocals(false)
-            .inheritInheritableThreadLocals(false)
-            .name("ttl-test-virtual")
-            .unstarted(() -> ttlOps.accept(ttl));
-        thread.start();
-        try {
-            thread.join();
-        } catch (InterruptedException e) {
-            throw new RuntimeException(e);
-        }
+        var carrierRef = new AtomicReference<Thread>();
 
-        if (!terminatedValues.isEmpty()) {
-            throw new AssertionError("Unexpected terminated values after virtual thread.join(): " +
-                                     terminatedValues);
-        }
+        // use a single worker thread pool as the "scheduler"
+        try (var pool = Executors.newSingleThreadExecutor()) {
 
-        // we now unblock the blocker thread but keep it running
-        synchronized (lock) {
-            lock.unblock = true;
-            lock.notify();
-        }
+            // capture carrier Thread
+            pool.submit(() -> carrierRef.set(Thread.currentThread()));
 
-        // carrier thread pool has a 30 second keep-alive time to terminate excessive carrier
-        // threads. Since blockerThread is still pinning one of them we hope for the other
-        // thread to be terminated...
-        try {
-            TimeUnit.SECONDS.sleep(31);
-        } catch (InterruptedException e) {
-            throw new RuntimeException(e);
+            ThreadFactory factory = virtualThreadBuilder(pool)
+                    .name("ttl-test-virtual-", 0)
+                    .allowSetThreadLocals(false)
+                    .factory();
+            try (var executor = Executors.newThreadPerTaskExecutor(factory)) {
+                executor.submit(() -> ttlOps.accept(ttl)).get();
+            }
+
+            if (!terminatedValues.isEmpty()) {
+                throw new AssertionError("Unexpected terminated values after virtual thread terminated: " +
+                                         terminatedValues);
+            }
         }
 
+        // wait for carrier to terminate
+        Thread carrier = carrierRef.get();
+        carrier.join();
```

Comment:
...would `close()`-ing the `pool` already guarantee that all pool thread(s) are already stopped? `join()` is not needed then. But it doesn't hurt.

Reply:
The close waits for the all tasks to finish. There is a small window between task completion and the thread termination.